### PR TITLE
Make at_exit and Printexc.register_printer thread-safe

### PR DESCRIPTION
**Pull request by gadmm**
This PR lives on top of #9570. (In particular it is best read per-commit.)

The first proper commit fixes data races in `at_exit` and `Printexc.register_printer` under systhreads using respectively a manual implementation and the new atomics. Using atomics for `at_exit` would cause dependency issues.

The second proper commit makes `Atomic` a dependency of `Stdlib` to implement `at_exit` with atomics. It took me _a while_ to manage to make it compile, and it seemed that I needed to redo the bootstrap, so please tell me in case the bootstrap is not supposed to be needed, I may have further issues with the build process.

### CODE REVIEW

**Comment by gadmm** [comment]
Note: I edited the `stdlib/dune` file. But `dune build @world` fails, which is already the case in trunk.

**Comment by gasche** [comment]
I propose to integrate the dependency change (having `atomic` linked before `stdlib`) in my PR so that you don't have to here.

(For Dune I believe that only `dune build @libs` is actively maintained.)

**Comment by gadmm** [comment]
That would be nice, feel free to cherry-pick from my PR.

**Comment by gasche** [comment]
Thinking more about it:
- I'm surprised by how painful it is to make the build system happy with your change.
- The Atomic PR is already dealing with a lot of boilerplate, so maybe piling the two together is not so nice. I would rather do a separate PR just for this.
- I think we should have a CamlinternalAtomic that Stdlib depends on, and have Atomic re-export its implementation, instead of using Stdlib__atomic in an awkward way in Stdlib.

**Comment by gadmm** [comment]
> * I'm surprised by how painful it is to make the build system happy with your change.

Indeed :)

> * The Atomic PR is already dealing with a lot of boilerplate, so maybe piling the two together is not so nice. I would rather do a separate PR just for this.

Would you or would I? Feel free if you want to open a separate PR if you would like to.

> * I think we should have a CamlinternalAtomic that Stdlib depends on, and have Atomic re-export its implementation, instead of using Stdlib__atomic in an awkward way in Stdlib.

I tried both versions. The CamlinternalAtomic version did not bring any simplification build-wise (I thought it would regarding the renaming business). Then it is a matter of taste, what you call "awkward" (where I do not see an issue). One advantage with my version is that it is a template people can reuse next time they need to make stdlib dependent on a module (without having to create a proxy module).

**Comment by gasche** [comment]
Yes, but it's also nice to say that stdlib should only depend on internal modules. (It's also a template people can follow.)

**Comment by gadmm** [comment]
Well, that is not obviously the meaning of Camlinternal...
* Only one Camlinternal is a dependency of Stdlib.
* It is used for something the user should not access because not subject to backwards-compat guarantees.

Here, nothing would be "internal" in CamlinternalAtomic. I really do not see the point in adding the complication (two more files). It is just a matter of getting used to seeing Stdlib__ in `stdlib.ml`.

**Comment by gasche** [comment]
Those are reasonable points, so maybe your PR could stay as it is. (But I can try to push a self-contained bootstrap to trunk after #9570 is merged, hoping that you don't have to do it in your PR anymore.)

**Comment by gadmm** [comment]
If there is nothing to change, I am fine with keeping it in this PR.

**Comment by xavierleroy** [comment]
> But I can try to push a self-contained bootstrap to trunk after #9570 is merged

It is not obvious at all to me why a bootstrap is needed.  And the fewer bootstraps we push, the better.  Does anyone understand why a bootstrap is needed?

**Comment by gadmm** [comment]
False alarm, no bootstrap is needed. It was just a collateral victim of my fight with the build system.

I will fix the remaining tests once I can rebase on top of the final version of #9570.

**Comment by gadmm** [comment]
Rebased on top of trunk.

**Review by gasche — APPROVED** [review]
I reviewed the code and I believe it is correct.

I have two questions:

- Are there any expected benefits with the current sequential runtime? (Is there a situation where those functions could call themselves in a non-reentrant way?)

- The implementation of `at_exit` is perplexing to me because it seems unnecessarily complicated. Why build a linked list of closures when we could have a plain list of exit callbacks? Why are we never trying to remove the exit functions that  we know have executed, instead of keeping them as no-ops forever?

The second question may very well be outside the scope of the present PR (which is keeping the previous logic, but thread-safe), but I ask myself whenever I encounter this piece of code, and I thought now would be an occasion to discuss it with exit experts.

**Comment by gadmm** [comment]
> I reviewed the code

Thanks!

>  * Are there any expected benefits with the current sequential runtime? (Is there a situation where those functions could call themselves in a non-reentrant way?)

Yes, there are races with systhread if the tick falls on the cons/the closure allocation. Then concurrently-installed at_exit handlers and exception printers can be overwritten.

>  * The implementation of `at_exit` is perplexing to me because it seems unnecessarily complicated. Why build a linked list of closures when we could have a plain list of exit callbacks? Why are we never trying to remove the exit functions that  we know have executed, instead of keeping them as no-ops forever?

Would that necessarily be simpler? At some point I did read all the issues and PRs associated with `at_exit` (https://github.com/ocaml/ocaml/issues?q=at_exit) and I wish I had the brain availability to go through all of it again, but I remember that alternative implementations were tried and the most straightforward was chosen. In the event that at_exit is changed to a sink that does not return, removing handlers would not matter (currently, if it does matter for a user then they make an unintended use of at_exit).

That being said, it would indeed be nice to have a place to discuss possible improvements to at_exit, but can I redirect the discussion to the issue #7253 instead? it might provide better context for the discussion.

**Comment by gadmm** [comment]
The build system has not said its last word yet. CheckDepend complains about `stdlib/.depend`, but running `make alldepend` only updates `debugger/.depend` on my side (the PR already updates `stdlib/.depend`).

**Comment by gadmm** [comment]
> * Are there any expected benefits with the current sequential runtime?

Implicit to your question is the idea that this would make it compatible with multicore, right? In fact that was not the goal, and in the case of `at_exit` one would need at least some synchronisation of the running flag, and maybe even rethink the semantics of `at_exit` to make sure that a handler runs in its original domain.

File: stdlib/.depend
Comment (gasche) [inline stdlib/.depend]: @gadmm those are the lines that CheckDepend complains about, I believe. I will try to improve the check to show the difference.

Comment (gadmm) [reply stdlib/.depend]: Well-spotted. So you have proof that I _did_ try with CamlinternalAtomic ;) Their source were committed at some point so I must have done some bad manoeuvre with git. Showing the diff in the test would definitely have helped, thanks.

**Comment by gadmm** [comment]
Following your question, I have pushed a third commit that implements synchronisation for the flag in the way it was intended for systhread at #1790 but using atomics instead.

**Comment by gasche** [comment]
This seems good to merge if the CI agrees.

Note: you wouldn't have to use `Stdlib__atomic` at all if the module-aliases block below was moved to the top of the file. I don't know why we are not doing that (first the aliases, and then the Pervasives code).

**Comment by gadmm** [comment]
@xavierleroy mentions that this (probably) breaks dune build. Xavier, what is the command line to reproduce?

**Comment by xavierleroy** [comment]
The CI script is in the repository: `tools/ci/inria/dune-build`.  It should reproduce the error. Cc @trefis .

**Comment by gadmm** [comment]
This must have to do with the dune option `modules_before_stdlib`. Is it compatible with the `stdlib__` prefixing trick? I could not determine this, since here `dune build @libs` is broken both before and after the second commit (though the second one adds an error about `Stdlib__atomic`). Any help welcome.

**Comment by gadmm** [comment]
I will wait to see if someone chimes in; if nobody does do not hesitate to ping me if you want me to fix it by going back to `CamlinternalAtomic`.

**Pull request merged**
